postgres: graphical query builder

svenklemm opened the pull request:
This enhances the postgres datasource frontend to add a graphical query builder. This still needs a few more improvements before it can be merged but I'm posting this to get some early feedback/testing. 

#10095

TODO:
- [x] unify quoting
- [x] handle unix timestamp as time column (requires #12891)

![image](https://user-images.githubusercontent.com/31455525/43713067-9e7eb3f6-9978-11e8-9783-fa20aa5796d8.png)

svenklemm committed:
start query builder ui

svenklemm committed:
call render in query

svenklemm committed:
more query builder components

svenklemm committed:
query builder changes

svenklemm committed:
add postgres_query.ts

svenklemm committed:
enhance render function

svenklemm committed:
add groupby to querybuilder
remove unused aggregations

svenklemm committed:
add metric column selector

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
make metricColumn functional

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
add query_builder

svenklemm committed:
add where constraint handling

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
rename field to column

svenklemm committed:
fix variable interpolation

svenklemm committed:
clean up aggregation functions

svenklemm committed:
fix group by column

svenklemm committed:
set rawSQL when rendering query builder query

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
use metricColumn in query builder

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
quote schema and table

svenklemm committed:
properly quote where constraint parts

svenklemm committed:
handle variables in where constraints

svenklemm committed:
remove limit

svenklemm committed:
cleanup where segment handling

svenklemm committed:
remove unused setting

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
add regex operators

svenklemm committed:
dont quote where constraints

svenklemm committed:
put values for IN in parens

svenklemm committed:
revert special handling for IN

svenklemm committed:
join multivalue variables with ,

svenklemm committed:
properly handle IN queries

svenklemm committed:
add buildAggregateQuery

svenklemm committed:
read aggregate functions from database

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

daniellee commented:
Looking good! Here's my feedback so far (you probably know all of this already :wink:):

## Where Clause

For the where clause, at the moment you have to manually quote the value for string values:

![image](https://user-images.githubusercontent.com/434655/37672580-9731b142-2c6e-11e8-9deb-9c21e59e0633.png)

I saw that Clickhouse had the same thing for their adhoc filters but elasticsearch and InfluxDB filters figure that out automatically. It is currently a bit confusing with the dropdown as the strings are not quoted:

![image](https://user-images.githubusercontent.com/434655/37672707-d4192392-2c6e-11e8-9522-fadc2592ef1d.png)

What are your thoughts on this?

## Group by Time and Metric

Couldn't get this query to work.
![image](https://user-images.githubusercontent.com/434655/37673049-936aaaf4-2c6f-11e8-992d-3c68f26c926c.png)

When I toggled to raw mode (btw totally amazing that you can toggle to edit mode and keep the query!), the query is slightly wrong:

`SELECT $__timeGroup("createdAt",1m),"value","hostname" AS metric FROM "public"."grafana_metric" WHERE $__timeFilter("createdAt") GROUP BY 1 ORDER BY 1`

Think the group by should be `GROUP BY 1,3`.

## Metric Column field

The dropdown didn't show a list of possible columns to use but it worked with manually entering the column name.

## Removing Stuff

Can't remove a metric name or group bys (except for the first group by).

Can remove `time` but not hostname:
![image](https://user-images.githubusercontent.com/434655/37673324-33f5be8c-2c70-11e8-9d76-24f2cb9a0c60.png)

Will test more but just to say it again - great work Sven!

torkelo commented:
This looks (functionally) really great! 

I wish this was not needed as this duplicates one of the more complex and messy parts of of Grafana (InfluxDB query editor), but I do see the value in this for making SQL queries more accessible and easier for people to write.  

I wonder if there is a better approach than using InfluxDB query editor as a reference, one that also supported some form of sub query, but that might be too complicated. Maybe break out a component for the where clause, that component and how it works and is built of metric segments is so messy. 

This will need a lot more work to be mergable as there is a ton of complex code here with zero tests. 

Question: where does there model > sql rendering happen? I can only find that it happens live when frontend queries the data source (in datasource.query method), but that means alerting will not work or will have to duplicate all the model > rendering code (something that was unfortunate for InfluxDB and I want to avoid in future, that is reducing duplication between frontend and backend data source code)

svenklemm commented:
@daniellee 
Regarding the where clause i think the dropdown suggestions should be returned quoted already so they work out of the box. I think autoquoting where parts will limit flexibility for more advanced use cases.

I'll fix the other stuff you mentioned once i get the aggregate function stuff handled properly. I didnt get it to work yet even with your help. The aggregate functions are fetched and if you add a 2nd query you will see aggregate functions actually available in the database. But maybe 2 levels for the SELECT part is overkill anyway since only the aggregate submenu would have multiple entries.

@torkelo 
There is no other editor that could be used as reference though, is there? 

If the code sets rawSQL on every render call then render only needs to be called when you are actually editing the query and alerts should continue to work. No backend changes are required for this to work.

I'll write tests once I get the aggregation function stuff working properly.

svenklemm committed:
support non-nested menu entries

svenklemm committed:
handle aggregate functions more generic

svenklemm committed:
remove unused import

svenklemm committed:
remove categories from queryPart

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

TSheahan commented:
This pull request makes for a very interesting read, and my organization could really use the feature!

Could anyone clarify what this means, in terms of how things are tracking towards the release of this feature?

daniellee commented:
@TSheahan If this PR is close to mergeable then we add it to the milestone so that someone from the core team reviews it and merges it. But it is ready when it is ready so it might get moved to the next milestone.

daniellee commented:
@svenklemm let me know when I should review this again or if you have gotten stuck with some frontend stuff.

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm commented:
@daniellee unfortunately i was busy with other stuff so i didnt get to work on this very much. I started working on a component similar to query_part to handle all the elements of the query but i havent gotten very far with this yet.

svenklemm committed:
return values quotes for suggestions in where expression

svenklemm committed:
quote column name in buildValueQuery

svenklemm committed:
add sql_part component

svenklemm committed:
use sql part component

svenklemm committed:
use sql part component

svenklemm committed:
separate label in template from type

svenklemm committed:
fix metric column when using group by

svenklemm committed:
make param wrapper configurable

svenklemm committed:
set style for registered query components

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
fix error message

svenklemm committed:
make separator configurable

svenklemm committed:
send param in callback for get-param-options

svenklemm committed:
remove unused function renderAdhocFilters

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
use sqlPart for ui parts

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
mv query_part to sql_part

svenklemm committed:
rearrange elements of query builder

svenklemm committed:
include where constraints in query generation

svenklemm committed:
code formatting

svenklemm committed:
remove dead code from sql_part
fix where clause query generation

svenklemm committed:
fix where constraint handling

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
fix group by ui

svenklemm committed:
do not autoquote identifiers

svenklemm committed:
dont autoquote, suggest quoted values if requried

svenklemm committed:
fix constraint removal

svenklemm committed:
remove dead code, make label more flexible

svenklemm committed:
remove hardcoded $__timeFilter, make macros functional in where clause

svenklemm committed:
fix where clause generation

svenklemm committed:
put updateParam back in

svenklemm committed:
remove unused function removeSelect

svenklemm committed:
rename quoteLiteral to quoteIdentAsLiteral

svenklemm committed:
refactor transformToSegments

svenklemm committed:
dont break on panels that dont have rawQuery set

svenklemm committed:
handle pre query gui queries gracefully

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
add None to metric column suggestions

svenklemm committed:
formatting

svenklemm committed:
show current value in dropdown when its not part of list

svenklemm committed:
Revert "show current value in dropdown when its not part of list"

This reverts commit 76275cae373a4114afdd2086f0f39968d5d99585.

svenklemm committed:
add current value to dropdown if its not in resultset

svenklemm committed:
fix spelling

svenklemm committed:
rename inputBlur to switchToLink

svenklemm committed:
fix editing expression parts

svenklemm committed:
fix enter in sql_part_editor

svenklemm committed:
add column alias when add aggregate function

svenklemm committed:
remove superfluous braces

svenklemm committed:
add aggregates when adding group by

svenklemm committed:
add alias when adding group by

svenklemm committed:
dont throw exception for unknown types

svenklemm committed:
set query gui as default handle old panels gracefully

svenklemm committed:
prepare sql part for special functions

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
Refactor render function on PostgresQuery

svenklemm committed:
Refactor time column sql generation

svenklemm committed:
refactor SQL generation for value columns

svenklemm committed:
Refactor where clause generation

svenklemm committed:
Refactor group by query generation

svenklemm committed:
Refactor metric column sql generation

svenklemm committed:
Refactor value column SQL generation

svenklemm committed:
add rate and increase special functions

svenklemm committed:
partition by metricColumn when using increase

svenklemm committed:
handle counter overflow and resets in rate

svenklemm committed:
indent generated SQL

svenklemm committed:
send timerange with metricFindQuery

svenklemm committed:
fix tests for postgres datasource

svenklemm committed:
add tests for query generation

svenklemm committed:
Add templateSrv to PostgresQuery tests

svenklemm committed:
use let for variable declaration

svenklemm committed:
refactor addGroupBy and removeGroupByPart

svenklemm committed:
refactor PostgresQueryCtrl and PostgresQuery

svenklemm committed:
refactor PostgresQueryCtrl

svenklemm committed:
refactor PostgresQuery

svenklemm committed:
fix bug in query generation with metricColumn

svenklemm committed:
remove render code from sql_part

svenklemm committed:
refactor adding sqlPart

svenklemm committed:
remove unused import

svenklemm committed:
fix rate special function when using group by

svenklemm committed:
filter datatype for groupby suggestions

svenklemm committed:
dont expand variables in rawSql

svenklemm committed:
rename PostgresQueryBuilder to PostgresMetaQuery

svenklemm committed:
add popover for metric column

svenklemm committed:
add query to find metric table

svenklemm committed:
dont run queries if target has no table set

svenklemm committed:
fix test for query generation

svenklemm committed:
add groupby when adding first aggregate

svenklemm commented:
@torkelo  I've written a component to handle the query parts that can't be handled by a single MetricSegment. I hope this addresses the issues you found with the initial version.
@daniellee I think now is a good time for another look. 
I also have a docker-image of this branch at svenklemm/grafana-dev

svenklemm committed:
fix pre gui queries shortcircuit

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=11081) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=11081) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: svenklemm<br/>:x: Sven Klemm<br/><hr/>**Sven Klemm** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=11081) it.</sub>

svenklemm committed:
set explicit order for rate and increase

svenklemm committed:
add order by to metadata queries

daniellee commented:
@svenklemm I just did a quick test using the postgres docker block included in Grafana and I'm getting errors for the aggregate dropdowns and for the metric name column.

Error: `pq: column "aggkind" does not exist` for this query:

`SELECT DISTINCT proname FROM pg_aggregate INNER JOIN pg_proc ON pg_aggregate.aggfnoid = pg_proc.oid INNER JOIN pg_type ON pg_type.oid=pg_proc.prorettype WHERE pronargs=1 AND typname IN ('float8') AND aggkind='n' ORDER BY 1`

Am I doing something wrong? Or does this query not work for certain versions of postgres?

svenklemm commented:
@daniellee I developed this against postgres 10 and didnt check for compatibility with old versions yet. That aggregate query needs 9.4 I think while the docker block in grafana uses 9.3. I'll do a pass over these metadata queries to make sure they work with older versions aswell. For testing this its probably best to use postgres 9.6 or 10 for now.

svenklemm committed:
fix window function query without group by

svenklemm committed:
remove tableschema from query builder ui

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
refactor schema query generation

svenklemm committed:
consistent nameing fro group and select

svenklemm committed:
refactor column function handling

svenklemm committed:
refactor function handling in query builder

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
add first and last support

svenklemm committed:
rename special to windows

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm commented:
@daniellee The offending query for aggregate functions is removed now, what problems did you have with the metric column?

svenklemm committed:
Merge branch 'master' into postgres-query-builder

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
add more prominent button for switching edit mode

svenklemm committed:
add warning when switching from raw sql mode

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
Merge remote-tracking branch 'upstream/master' into postgres-query-builder

svenklemm committed:
use $__timeGroupAlias macro

svenklemm committed:
adjust frontend test

svenklemm committed:
add moving average to query builder

svenklemm committed:
dont order for aggregate

svenklemm committed:
fix suggestion query

svenklemm committed:
dont break default parameters for functions

svenklemm committed:
Merge branch 'master' into postgres-query-builder

svenklemm committed:
add previous fill mode to query builder

svenklemm committed:
Merge branch 'master' into postgres-query-builder

svenklemm committed:
fix datatype query

svenklemm committed:
get timecolumn datatype on timecolumn change

svenklemm committed:
Merge branch 'master' into postgres-query-builder

svenklemm committed:
Merge branch 'master' into postgres-query-builder

svenklemm committed:
Merge branch 'master' into postgres-query-builder

svenklemm committed:
keep jsonData in PostgresDatasource

svenklemm committed:
only show first/last aggregate when timescaledb is enabled

svenklemm committed:
add postgres version to datasource config

svenklemm committed:
require postgres 9.4+ for ordered set aggregate functions

svenklemm committed:
use pointer cursor for buttons in query editor

svenklemm committed:
detect postgres version for saved datasources

svenklemm committed:
autodetect timescaledb when version >= 9.6

svenklemm committed:
fix variable escaping

svenklemm committed:
rename postgres_query.jest.ts to .test.ts

svenklemm committed:
Don't do value suggestions for numeric and timestamp

svenklemm committed:
persist datatype information

svenklemm committed:
update persisted parts on param change

svenklemm committed:
suggest operators depending on datatype

svenklemm committed:
Merge branch 'master' into postgres-query-builder

svenklemm committed:
Merge branch 'master' into postgres-query-builder

svenklemm committed:
use unixEpochGroup macro for unix timestamp

svenklemm committed:
make suggested filter macro depend on type

svenklemm committed:
fix timeColumnType assignment

svenklemm committed:
only allow 1 filter macro in where clause

svenklemm committed:
prefill editor with first metric table

svenklemm committed:
unify quoting

svenklemm committed:
fix tableSegment and timeColumnSegment after table suggestion

svenklemm committed:
Merge branch 'master' into postgres-query-builder

svenklemm committed:
Merge branch 'master' into postgres-query-builder

marefr commented on public/app/plugins/datasource/postgres/meta_query.ts:
```diff
@@ -0,0 +1,176 @@
+
+  getOperators(datatype: string) {
+    switch (datatype) {
+      case 'float4':
```
Having metadata returned from API with datatypes like string, numeric could help to get rid of checking specific postgres datatypes here. However, you still need to do it somewhere, but maybe certain parts of this then could be extracted to a common base class.

marefr commented on public/app/plugins/datasource/postgres/partials/config.html:
```diff
@@ -41,6 +41,17 @@ <h3 class="page-heading">PostgreSQL Connection</h3>
+		<span class="gf-form-label width-9">
+			Version
+			<info-popover mode="right-normal" position="top center">
+				This option controls what functions are available in the PostgreSQL query builder.
```
> This option controls what functions are available in the PostgreSQL qraphical query builder.

Maybe 
... available when using the PostgreSQL query builder.

marefr commented on public/app/plugins/datasource/postgres/partials/query.editor.html:
```diff
@@ -1,43 +1,141 @@
-			</code-editor>
-		</div>
-	</div>
+<query-editor-row query-ctrl="ctrl" has-text-edit-mode="true">
```
This file feels very generic for a SQL query editor and probably something we can create a component for later :+1: 

The exception is the help text, but should be trivial to fix.

marefr commented on public/app/plugins/datasource/postgres/postgres_query.ts:
```diff
@@ -0,0 +1,285 @@
+import _ from 'lodash';
```
This class should also be fairly easy to extract to common base class/component eventually :+1:

marefr commented on public/app/plugins/datasource/postgres/query_ctrl.ts:
```diff
@@ -1,12 +1,10 @@
 import _ from 'lodash';
+import appEvents from 'app/core/app_events';
```
Some parts from this file could also probably be extract to base class/component eventually :+1:

marefr commented on public/app/core/components/sql_part/sql_part.ts:
```diff
@@ -0,0 +1,74 @@
+import _ from 'lodash';
```
This component looks good from a "generic" code point of view :+1:

marefr reviewed (commented):
Added some comments.

**Problems:**
Trying out our pre-built development dashboards ([here](https://github.com/grafana/grafana/blob/master/devenv/dev-dashboards/datasource_tests_postgres_fakedata.json) and [here](https://github.com/grafana/grafana/blob/master/devenv/dev-dashboards/datasource_tests_postgres_unittest.json)) with postgres version 9.3 generates error for certain panels.

I'm guessing that if rawQuery=true the plugin should behave same way as in current master.

<details>
 <summary>Stacktrace</summary>
```
Panel data error: TypeError: Cannot read property 'escapeLiteral' of undefined
    at push../public/app/plugins/datasource/postgres/postgres_query.ts.PostgresQuery.quoteLiteral (postgres_query.ts:51)
    at arrayMap (lodash.js:639)
    at Function.map (lodash.js:9554)
    at PostgresQuery.push../public/app/plugins/datasource/postgres/postgres_query.ts.PostgresQuery.interpolateQueryStr (postgres_query.ts:76)
    at TemplateSrv.push../public/app/features/templating/template_srv.ts.TemplateSrv.formatValue (template_srv.ts:91)
    at template_srv.ts:216
    at String.replace (<anonymous>)
    at TemplateSrv.push../public/app/features/templating/template_srv.ts.TemplateSrv.replace (template_srv.ts:188)
    at PostgresQuery.push../public/app/plugins/datasource/postgres/postgres_query.ts.PostgresQuery.render (postgres_query.ts:93)
    at datasource.ts:51
```
</details>

marefr commented on public/app/plugins/datasource/postgres/partials/config.html:
```diff
@@ -41,6 +41,17 @@ <h3 class="page-heading">PostgreSQL Connection</h3>
+		</span>
+	</div>
 	<div class="gf-form">
 		<gf-form-switch class="gf-form" label="TimescaleDB" tooltip="Use TimescaleDB features (e.g., time_bucket) in Grafana" label-class="width-9" checked="ctrl.current.jsonData.timescaledb" switch-class="max-width-6"></gf-form-switch>
```
For a new user that don't know what TimescaleDB is I think the description could be better. 

Maybe use similar approach to http datasources which have a Help button, see [code](https://github.com/grafana/grafana/blob/master/public/app/features/plugins/partials/ds_http_settings.html#L32). 

A short description with a link to TimescaleDB github repository would be nice I think.

svenklemm committed:
add min interval to postgres datasource

svenklemm commented on public/app/plugins/datasource/postgres/config_ctrl.ts:
```diff
@@ -0,0 +1,57 @@
+
+  // the value portion is derived from postgres server_version_num/100
+  postgresVersions = [
+    { name: '9.3', value: 903 },
```
9.4 is the first version that adds a new feature (ordered set aggregate functions), so anything below 9.4 would be treated the same
the number is  how you can query it from postgres itself (```SELECT current_setting('server_version_num');``` divided by 100, the version being integer also makes version checks easier

marefr replied:
Yes makes sense now, thanks.

svenklemm committed:
fix handling of variable interpolation for IN expresions

svenklemm committed:
Merge branch 'master' into postgres-query-builder

svenklemm committed:
improve description for timescaledb option

svenklemm committed:
make default mode for table panels raw editor

marefr commented on public/app/plugins/datasource/postgres/plugin.json:
```diff
@@ -18,5 +18,10 @@
-  "metrics": true
+  "metrics": true,
+
+  "queryOptions": {
```
Don't think this is needed right now. The $__interval should work anyway. Think it's for calculating intervals on backend.

marefr commented on public/app/plugins/datasource/postgres/partials/config.html:
```diff
@@ -38,6 +38,18 @@ <h3 class="page-heading">PostgreSQL Connection</h3>
 	</div>
 </div>
 
+<div class="gf-form-group">
```
This is not used right now so think we can skip this for now. The $__interval should work anyway. Think it's for calculating intervals on backend.

marefr reviewed (changes requested):
Editor still seems a bit buggy regarding columns wrapped in quotes and needs to be a bit clever to be ably to support users in a user friendly way.

1. I create a new query using gdev-postgres-ds-tests and get the following fields prefilled:
![image](https://user-images.githubusercontent.com/1668778/44743447-d2d10c80-ab02-11e8-8286-587bdf1992f5.png)

2. I select table=metric_values and I would expect the time column to automatically change to a time column available in that table. Instead column from initial table are shown and generates an error "pq: column "c10_timestamp" does not exist". Also the column selected is still the same as for the initial table - would be nice if the columns was either removed and that a query not was sent to backend until every selection is "valid" or either change selected columns to a column from the current selected table.
3. I select time column="timeInt32". Would expect the $__timeFilter to change to $__unixEpochFilter, but it doesn't. This results in a postgres error "pq: invalid input syntax for integer: "2018-03-15T12:30:00Z"" which is not very clear why that is happening.
4. I manually removes $__timeFilter and adds $__unixEpochFilter. Query works
5. I add a group by time. Query fails again "pq: syntax error at or near "timeInt32"" which is not very clear why that is happening, see screenshot below. Seems like the alias incorrectly gets quote added.

![image](https://user-images.githubusercontent.com/1668778/44743177-0f503880-ab02-11e8-84ca-b80709f2848c.png)

6. Remove quotes and the query works and gives a plotted graph.

svenklemm committed:
update filter macro on time column change

svenklemm committed:
recheck timecolumn when changing table

svenklemm committed:
link to github instead

svenklemm committed:
handle quoting properly for table suggestion

svenklemm committed:
strip quotes when auto adding alias

svenklemm committed:
fix quoting

svenklemm commented on public/app/plugins/datasource/postgres/query_ctrl.ts:
```diff
@@ -54,6 +54,7 @@ export class PostgresQueryCtrl extends QueryCtrl {
       if (this.panelCtrl.panel.type === 'table') {
         this.target.format = 'table';
         this.target.rawSql = 'SELECT 1';
+        this.target.rawQuery = true;
```
Yes query builder expects you to have at least a time column and multiple value columns so i guess it should not be default. Queries it creates will still work but will be in the form of
```SELECT time, value1,value2,... FROM table``` so no wildcards and time is not optional.

marefr replied:
You're correct. Please ignore my earlier comment. No changes needed regarding this.

svenklemm committed:
remove unneeded queryOptions

svenklemm committed:
remove min time interval from datasource config

svenklemm committed:
Merge branch 'master' into postgres-query-builder

svenklemm committed:
check for correct quoting of multiple singlequotes

svenklemm committed:
hide Query Builder button for table panels

svenklemm committed:
document postgres version and TimescaleDB option

marefr commented:
Needs fixes:
1. Metric column suggestion varying varchar data type doesn't show up
2. Handle selected column when changing table

svenklemm committed:
fix suggestions for metric column

svenklemm committed:
fix metric column suggestions

svenklemm committed:
when changing table reset columns

svenklemm committed:
set default for timeGroup in query builder to $__interval

svenklemm committed:
fix timecolumn handling when table changes

marefr committed:
when changing table, refresh panel once after columns have been changed

svenklemm committed:
fix timeFilter resetting when changing table

svenklemm committed:
reset metric column when changing table

svenklemm committed:
reset metric column when changing table

svenklemm committed:
Merge branch 'postgres-query-builder' of git+ssh://github.com/svenklemm/grafana into postgres-query-builder

svenklemm committed:
Merge branch 'master' into postgres-query-builder

svenklemm committed:
Merge branch 'master' into postgres-query-builder

svenklemm committed:
Merge branch 'master' into postgres-query-builder

marefr committed:
docs: postgres provisioning
